server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,193 @@
+        MasterServiceTaskQueue<NodeTransportVersionTask> taskQueue,
+        ClusterAdminClient client,
+        Scheduler scheduler
+    ) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+        this.scheduler = scheduler;
+    }
+
+    class NodeTransportVersionTask implements ClusterStateTaskListener {
+        private final Map<String, TransportVersion> results;
+        private final int retryNum;
+
+        NodeTransportVersionTask(Map<String, TransportVersion> results, int retryNum) {
+            this.results = results;
+            this.retryNum = retryNum;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for nodes {} to cluster state", results.keySet(), e);
+            scheduleRetry(results.keySet(), retryNum);
+        }
+
+        public Map<String, TransportVersion> results() {
+            return results;
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    // this node's transport version might have been updated already/node has gone away
+                    if (Objects.equals(builder.transportVersions().get(e.getKey()), INFERRED_TRANSPORT_VERSION)) {
+                        builder.putTransportVersion(e.getKey(), e.getValue());
+                    }
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster() == false) return; // only if we're master
+
+        // if the min node version > 8.8.0, and the cluster state has some transport versions == 8.8.0,
+        // then refresh all inferred transport versions to their real versions
+        // now that everything should understand cluster state with transport versions
+        if (event.state().nodes().getMinNodeVersion().after(Version.V_8_8_0)
+            && event.state().getMinTransportVersion().equals(INFERRED_TRANSPORT_VERSION)) {
+
+            // find all the relevant nodes
+            Set<String> nodes = event.state()
+                .transportVersions()
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue().equals(INFERRED_TRANSPORT_VERSION))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toSet());
+
+            updateTransportVersions(nodes, 0);
+        }
+    }
+
+    private void scheduleRetry(Set<String> nodes, int thisRetryNum) {
+        if (thisRetryNum >= MAX_RETRIES) {
```

Comment:
I'd rather we kept going forever, making noise in the logs if something is this badly broken. Otherwise there's just this one log message which may be days old and rolled out of the current logs.

However, if a node leaves the cluster then we should stop trying to get its transport version straight away. It looks to me as if we keep on trying with such nodes?

Reply:
Is there a specific exception that is thrown when you're getting node info for a node that doesn't exist anymore?

Reply:
I think it might just be treated as if it doesn't exist - see `org.elasticsearch.cluster.node.DiscoveryNodes#resolveNodes`

Reply:
So that's fine then, it'll not return anything about that node in the response, so that node will just be forgotten about.

Reply:
Ah ok I see that now. Sorry for the noise.